core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -252,4 +282,28 @@ public void setSource(Cache<K, V> source) {
     }
   }
 
+  private static class EventThreadLocal {
+    private final ThreadLocal<List<CacheEventWrapper>> threadLocal = new ThreadLocal<List<CacheEventWrapper>>() {
+      @Override
+      protected List<CacheEventWrapper> initialValue() {
+        return new ArrayList<CacheEventWrapper>();
+      }
+    };
+
+    public List<CacheEventWrapper> get() {
+      HashMap<Object, CacheEventWrapper> cacheEventMap = new HashMap<Object, CacheEventWrapper>();
+      for (CacheEventWrapper cacheEventWrapper : threadLocal.get()) {
+        if((cacheEventWrapper.cacheEvent.getType() == EventType.CREATED)
```

Comment:
This is here to cater to the scenario where a key ends up having more than one `CacheEvent`(`CREATE`, `UPDATE`, `REMOVE`). With each new event added to the thread local we make sure that if it is a `CacheEvent`, the key for that event does not have any other `CacheEvent` prior to it. If it does the older `CacheEvent` is marked `Failed`.